### PR TITLE
Indicate null/missing client certificate

### PR DESCRIPTION
**Pull request by anrossi**
## Description

When `QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION` or `QUIC_CREDENTIAL_FLAG_DEFER_VALIDATION` are present, and `QUIC_CREDENTIAL_FLAG_INDICATE_PEER_CERTIFICATE_RECEIVED` is specified, indicate a null certificate message to the app when no certificate was received from the peer.

## Testing

A new test was added to test this case when the client doesn't send a certificate. 

## Documentation
No documentation was added/updated.

Related items: #2573 #1953

### CODE REVIEW

File: src/inc/msquic.h [hunk 1]
```diff
@@ -110,6 +110,7 @@ typedef enum QUIC_CREDENTIAL_FLAGS {
     QUIC_CREDENTIAL_FLAG_SET_ALLOWED_CIPHER_SUITES              = 0x00002000,
     QUIC_CREDENTIAL_FLAG_USE_PORTABLE_CERTIFICATES              = 0x00004000,
     QUIC_CREDENTIAL_FLAG_USE_SUPPLIED_CREDENTIALS               = 0x00008000, // Schannel only currently
+    QUIC_CREDENTIAL_FLAG_INDICATE_NULL_CLIENT_CERTIFICATE       = 0x00010000,
```
Comment (nibanks) [inline src/inc/msquic.h:113]:
```suggestion
    QUIC_CREDENTIAL_FLAG_INDICATE_NULL_CERTIFICATE                    = 0x00010000,
```
I think we don't need to include CLIENT in the name. BTW, we need to add validation logic to fail with invalid parameter if this is passed on client.

File: src/platform/tls_openssl.c [hunk 2]
```diff
@@ -102,6 +102,11 @@ typedef struct CXPLAT_TLS {
     //
     BOOLEAN IsServer;
 
+    //
+    // Indicates if the peer sent a certificate.
+    //
+    BOOLEAN PeerCertReceived;
```
Comment (nibanks) [inline src/platform/tls_openssl.c:108]: Use a bit field for this and `IsServer` above.

**Review by nibanks — COMMENTED** [review]
Please add both positive and negative tests, for the new flag being present and missing.

File: src/platform/tls_openssl.c [hunk 3]
```diff
@@ -1964,6 +1972,16 @@ CxPlatTlsProcessData(
...
+        } else if ((TlsContext->SecConfig->Flags & QUIC_CREDENTIAL_FLAG_INDICATE_NULL_CLIENT_CERTIFICATE) &&
+            !TlsContext->PeerCertReceived) {
+            if (!TlsContext->SecConfig->Callbacks.CertificateReceived(
+                TlsContext->Connection,
+                NULL,
+                NULL,
+                0,
+                0)) {
+                // TODO kill the handshake
+            }
```
Comment (nibanks) [inline src/platform/tls_openssl.c:1984]:
```suggestion
            if (!TlsContext->SecConfig->Callbacks.CertificateReceived(
                  TlsContext->Connection,
                  NULL,
                  NULL,
                  0,
                  0)) {
                TlsContext->ResultFlags |= CXPLAT_TLS_RESULT_ERROR;
                goto Exit;
            }
```
I think this is what you want.

File: src/inc/msquic.h (hunk 1 above)
Comment (anrossi) [reply src/inc/msquic.h:113]: After discussion, removing this flag and making the existing flags work.

File: src/platform/tls_openssl.c [hunk 4]
```diff
@@ -1964,6 +1974,29 @@ CxPlatTlsProcessData(
...
+                TlsContext->SecConfig->Flags & QUIC_CREDENTIAL_FLAG_DEFER_CERTIFICATE_VALIDATION)) ?
+                    QUIC_STATUS_CERT_NO_CERT :
+                    QUIC_STATUS_SUCCESS;
+
+            if (!TlsContext->SecConfig->Callbacks.CertificateReceived(
+                TlsContext->Connection,
+                NULL,
+                NULL,
+                0,
+                ValidationResult)) {
```
Comment (nibanks) [inline src/platform/tls_openssl.c:1991]:
```suggestion
            if (!TlsContext->SecConfig->Callbacks.CertificateReceived(
                    TlsContext->Connection,
                    NULL,
                    NULL,
                    0,
                    ValidationResult)) {
```

File: src/platform/unittest/TlsTest.cpp [hunk 5]
```diff
@@ -1777,6 +1777,26 @@ TEST_F(TlsTest, ClientCertificateDeferValidation)
     DoHandshake(ServerContext, ClientContext);
 }
 
+TEST_F(TlsTest, ClientCertificateDeferValidationNoCert)
+{
+    CxPlatClientSecConfig ClientConfig(
+        QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION
+#ifdef QUIC_ENABLE_ANON_CLIENT_AUTH_TESTS
+        | QUIC_CREDENTIAL_FLAG_USE_SUPPLIED_CREDENTIALS
+#endif
```
Comment (nibanks) [inline src/platform/unittest/TlsTest.cpp:1786]: Why modify the behavior of a test with the `QUIC_ENABLE_ANON_CLIENT_AUTH_TESTS` and not have two separate tests?

File: CMakeLists.txt [hunk 6]
```diff
@@ -338,6 +338,8 @@ if(QUIC_TLS STREQUAL "schannel")
     list(APPEND QUIC_COMMON_DEFINES QUIC_DISABLE_0RTT_TESTS)
     message(STATUS "Disabling ChaCha20 support")
     list(APPEND QUIC_COMMON_DEFINES QUIC_DISABLE_CHACHA20_TESTS)
+    message(STATUS "Enabling anonymous client auth tests")
+    list(APPEND QUIC_COMMON_DEFINES QUIC_ENABLE_ANON_CLIENT_AUTH_TESTS)
```
Comment (nibanks) [inline CMakeLists.txt:342]: Why can't se support this behavior with OpenSSL?

File: src/platform/unittest/TlsTest.cpp (hunk 5 above)
Comment (anrossi) [reply src/platform/unittest/TlsTest.cpp:1786]: USE_SUPPLIED_CREDS is necessary for schannel, and is an schannel-only flag. I want this test to also cover OpenSSL.

File: CMakeLists.txt (hunk 6 above)
Comment (anrossi) [reply CMakeLists.txt:342]: We already do. USE_SUPPLIED_CREDS is a flag that is schannel-only. OpenSSL works just fine out of the box.

**Pull request merged**